### PR TITLE
Filing process routing - Phase 2 (Live data)

### LINKED ISSUE #404: [Navigation] Filing routing - Phase 2 - Live data
- [ ] Use Filing API to derive which steps are complete/incomplete

### PR DESCRIPTION
**Pull request by meissadia**
Closes #404  

## Changes
- [Filing] Automatically creates a new `Filing` when needed
- [Filing overview] Determines latest incomplete Filing step and links to it
- [Filing sub-pages] Creates placeholder pages for Errors, Warnings, Point of contact, and Complete
- [Filing sub-pages] Displays the status of all Steps at the top of each sub-page
- [Styling] Updates step label's font-weight and color based on the steps status (complete, incomplete)
- [Styling] Only displays a step as Complete (Dark blue bar, semibold font) when a user has navigated beyond the step in the Filing flow (i.e. When viewing `Warnings`, `Errors` would show complete)

## How to test this PR

1. Start fresh
2. [Landing page](http://localhost:8899/filing) - Verify all associated institutions are ready for Upload (which indicates that the Filing for this period has automatically been created)
3. Upload a ["failing"](https://github.com/cfpb/regtech-data-validator/tree/main/tests/data) data file.  Verify that you *cannot* navigate beyond the `Upload` page (`Save..` button disabled)
4. Upload a ["passing"](https://github.com/cfpb/regtech-data-validator/tree/main/tests/data) data file. Verify that you *can* navigate to the `Errors` page  (`Save..` button enabled). Also verify that the step indicator at the top of the page updates to reflect the status of upcoming steps. 
5. Return to the [Landing page](http://localhost:8899/filing), verify that the Institution that you have uploaded data for now links to the latest "Incomplete" step (probably "Point of Contact").

## Screenshots (Updated 4/25)
https://github.com/cfpb/sbl-frontend/assets/2592907/54280e03-a2e5-4a10-b0d2-5a4127bac342

### CODE REVIEW

File: src/pages/Filing/FilingApp/FilingContact.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { useParams } from 'react-router-dom';
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/pages/Filing/FilingApp/FilingContact.tsx:1]: @billhimmelsbach Create a separate ticket for this feature

https://github.com/cfpb/sbl-frontend/issues/416

**Review by billhimmelsbach — COMMENTED** [review]
I want to merge this in! Let's fix the git conflicts @meissadia, and I'll approve it. Natalia had some feedback from yesterday, but let's deal with that in a separate design review PR.

**Comment by meissadia** [comment]
@billhimmelsbach Refactoring complete.  Merge conflicts squashed. Ready for re-review.

File: src/api/requests/fetchFilingSubmissionLatest.ts [hunk 2]
```diff
@@ -93,7 +93,10 @@ async function retryRequestWithDelay(
 function shouldRetry(response: AxiosResponse<SubmissionResponse>): boolean {
   // Check if the response has a 'state' property equal to "VALIDATION_IN_PROGRESS"
   // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition, @typescript-eslint/no-unused-expressions, prettier/prettier
-  return (response.data?.state && response.data.state) === fileSubmissionState.VALIDATION_IN_PROGRESS;
+  return (
+    (response.data.state && response.data.state) ===
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/api/requests/fetchFilingSubmissionLatest.ts]: Accepted

File: src/utils/useFilingAndSubmissionInfo.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    queryFn: async (): Promise<FilingType> =>
+      createFiling(auth, lei, filingPeriod),
+    enabled:
+      !existingFiling.isLoading && !isObjectInitialized(existingFiling.data),
+    staleTime,
+  });
+
+  const filing = createdFiling.data ?? existingFiling.data;
+
+  const latestSubmission = useQuery({
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/utils/useFilingAndSubmissionInfo.tsx:60]: When replacing this call with `useGetSubmissionLatest(lei, filingPeriod)` I get stuck in an infinite `Loading` state.  Haven't figured out why, I'll need to explore more next week. 
 
For now I'm leaving this piece as is.

[the remark this answers is not in the document]
Comment (meissadia) [reply src/utils/useFilingAndSubmissionInfo.tsx:60]: Actually, even without that `useGetSubmissionLatest(lei, filingPeriod)` call I'm seeing the `/upload` page hang on initial load.  `initialGetSubmissionLatestFetched` seems buggy.  Again, will need to investigate more.

File: src/pages/Filing/FilingApp/FileSubmission.tsx [hunk 4]
```diff
@@ -79,10 +79,15 @@ export function FileSubmission(): JSX.Element {
...
+    if (!isFetchingGetSubmissionLatest && !errorGetSubmissionLatest) {
+      setInitialGetSubmissionLatestFetched(true);
       setDataGetSubmissionLatest(actualDataGetSubmissionLatest);
     }
-  }, [actualDataGetSubmissionLatest]);
+  }, [
+    actualDataGetSubmissionLatest,
+    isFetchingGetSubmissionLatest,
+    errorGetSubmissionLatest,
+  ]);
```
Comment (meissadia) [inline src/pages/Filing/FilingApp/FileSubmission.tsx:90]: @sigmund.tang 

It seems that with the check `if (actualDataGetSubmissionLatest)` you are trying to ensure that the call to `/submission/latest` is complete and successful, then setting some flags in the local state to remove the `Loading` indicator? 
 
However, for new Filers who have not uploaded any submissions yet, `actualDataGetSubmissionLatest` remains empty/falsy, so the `initialGetSubmissionLatestFetched` flag stays `false` and the `Loading` indicator remains displayed.

File: src/utils/useFilingAndSubmissionInfo.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (meissadia) [reply src/utils/useFilingAndSubmissionInfo.tsx]: Other loading issue resolved, I've updated this to use useGetSubmissionLatest hook 👍🏾

**Comment by meissadia** [comment]
@shindigira @billhimmelsbach Issue with infinite `loading` state has been resolved.  All other PR comments have been addressed.  
 
This is ready for one last review.

File: src/utils/useFilingAndSubmissionInfo.tsx [hunk 5]
```diff
@@ -39,34 +40,36 @@ export const useFilingAndSubmissionInfo = ({
...
+    filing,
     isLoading,
-    submission: isObjectInitialized(latestSubmission.data)
-      ? latestSubmission.data
-      : { state: 'SUBMISSION_STARTED', refetch: latestSubmission.refetch },
-    refetchFiling: existingFiling.refetch,
-    refetchSubmission: latestSubmission.refetch,
+    submission: isObjectInitialized(submission)
+      ? submission
+      : { state: 'SUBMISSION_STARTED', refetch },
```
[the remark this answers is not in the document]
Comment (meissadia) [reply src/utils/useFilingAndSubmissionInfo.tsx]: Removed.  That `refetch` bit was left over from my attempts to update the step indicator (completion status of future steps) after a successful action on the same page, but that design has been squashed.

**Review by billhimmelsbach — APPROVED** [review]
Looks good @meissadia! Thanks so much for your hard work on this one! 🥳

Chatted with @meissadia about consolidating the filing status definitions, which we can do in a later PR. 👍

**Pull request merged**